### PR TITLE
Add fault injection for persistence API

### PR DESCRIPTION
**Pull request by Ardagan**
**What changed?**
Add option to inject failures in persistence api.

**Why?**
Open path for extra testing scenarios.

**How did you test it?**
Manual/auto tests.

**Potential risks**
Instability in new integration tests.

**Is hotfix candidate?**
no

### CODE REVIEW

File: common/persistence/client/fault_injection.go [hunk 1]
```diff
@@ -0,0 +1,925 @@
...
+		return nil
+	}
+
+	if roll := p.r.Float64(); roll < p.rate {
+		msg := fmt.Sprintf("FaultInjection injected error with rate %f, roll: %f", p.rate, roll)
+		errTypeRoll := p.r.Float64()
+		if errTypeRoll < 0.1 {
+			return serviceerror.NewInternal(msg)
+		}
+		return &persistence.TimeoutError{Msg: msg}
```
Comment (wxing1292) [inline common/persistence/client/fault_injection.go:119]: can this be configurable? 
e.g. initialization logic will provide a list of errors to return, with corresponding distribution

File: common/persistence/client/fault_injection.go [hunk 2]
```diff
@@ -0,0 +1,925 @@
...
+	if err != nil {
+		return baseFactory, err
+	}
+	return NewFaultInjectionTaskStore(d.config, baseFactory)
+}
+
+func (d *FaultInjectionDataStoreFactory) NewShardStore() (persistence.ShardStore, error) {
+	baseFactory, err := d.baseFactory.NewShardStore()
+	if err != nil {
+		return baseFactory, err
```
Comment (wxing1292) [inline common/persistence/client/fault_injection.go:170]: return err?

File: common/persistence/client/fault_injection.go (hunk 1 above)
Comment (yiminc) [reply common/persistence/client/fault_injection.go:119]: @Ardagan I think we agreed to make error types (and their likely-hood ) configurable. It it planed in your follow up PRs?

File: common/persistence/client/fault_injection.go (hunk 2 above)
Comment (yiminc) [reply common/persistence/client/fault_injection.go:170]: should return nil, err

File: common/persistence/client/fault_injection.go [hunk 3]
```diff
@@ -0,0 +1,925 @@
...
+	}
+	return NewFaultInjectionShardStore(d.config, baseFactory)
+}
+
+func (d *FaultInjectionDataStoreFactory) NewMetadataStore() (persistence.MetadataStore, error) {
+	baseStore, err := d.baseFactory.NewMetadataStore()
+	if err != nil {
+		return baseStore, err
```
Comment (yiminc) [inline common/persistence/client/fault_injection.go:178]: return nil, err

Comment (yiminc) [reply common/persistence/client/fault_injection.go:178]: and same for all below functions.

File: common/persistence/client/fault_injection.go [hunk 4]
```diff
@@ -0,0 +1,925 @@
...
+}
+
+func (c *FaultInjectionClusterMetadataStore) PruneClusterMembership(request *persistence.PruneClusterMembershipRequest) error {
+	if err := c.errorGenerator.Generate(); err != nil {
+		return err
+	}
+	return c.baseCMStore.PruneClusterMembership(request)
+}
+
+func NewFaultInjectionMetadataStore(
```
Comment (yiminc) [inline common/persistence/client/fault_injection.go:693]: should the function return error, does not seem to be possible to return any error?

File: common/persistence/client/fault_injection.go (hunk 1 above)
Comment (yiminc) [reply common/persistence/client/fault_injection.go:119]: At least, we want to mimic a few other error types, like OwnershipLostError etc.

**Review by yiminc — APPROVED** [review]
overall looks good. Major comment is the ability to config the error types and their possibility.

File: common/persistence/client/fault_injection.go (hunk 1 above)
Comment (Ardagan) [reply common/persistence/client/fault_injection.go:119]: Seems that OwnershipLostError  mostly sums it up. Updating.

**Comment by Ardagan** [comment]
Synced with Yimin, got green light to merge.

**Pull request merged**